src/sentry/snuba/metrics/extraction.py

```diff
@@ -196,69 +231,169 @@ def query_hash(self) -> str:
-        # start and end parameters are required, but not used
-        params={"start": datetime.now(), "end": datetime.now()},
-    )
+    def convert(self) -> RuleCondition:
+        """
+        Converts the token stream into a rule condition.
+
+        This function can raise an exception if the token stream is structurally
+        invalid or contains fields that are not supported by the rule engine.
+        """
+
+        condition = self._expr()
+        if self._position < len(self._tokens):
+            raise ValueError("Unexpected trailing tokens")
+        return condition
+
+    def _peek(self) -> Optional[QueryToken]:
+        """Returns the next token without consuming it."""
+
+        if self._position < len(self._tokens):
+            return self._tokens[self._position]
+        else:
+            return None
+
+    def _consume(self, pattern: Union[str, Type[T]]) -> Optional[T]:
+        """
+        Consumes the next token if it matches the given pattern.
+
+        The pattern can be:
+         - a literal string, in which case the token must be equal to the string
+         - a type, in which case the token must be an instance of the type
+
+        Returns the token if it matches, or ``None`` otherwise.
+        """
+        token = self._peek()
+
+        if isinstance(pattern, str) and token != pattern:
+            return None
+        elif isinstance(pattern, type) and not isinstance(token, pattern):
+            return None
+
+        self._position += 1
+        return cast(T, token)
+
+    def _expr(self) -> RuleCondition:
+        terms = [self._term()]
+
+        while self._consume("OR") is not None:
+            terms.append(self._term())
+
+        if len(terms) == 1:
+            return terms[0]
+        else:
+            return {"op": "or", "inner": terms}
+
+    def _term(self) -> RuleCondition:
+        factors = [self._factor()]
+
+        while self._peek() not in ("OR", None):
+            self._consume("AND")  # AND is optional and implicit, ignore if present.
+            factors.append(self._factor())
+
+        if len(factors) == 1:
+            return factors[0]
+        else:
+            return {"op": "and", "inner": factors}
+
+    def _factor(self) -> RuleCondition:
+        if filt := self._consume(SearchFilter):
+            return self._filter(filt)
+        elif paren := self._consume(ParenExpression):
+            return SearchQueryConverter(paren.children).convert()
+        elif token := self._peek():
+            raise ValueError(f"Unexpected token {token}")
+        else:
+            raise ValueError("Unexpected end of query")
+
+    def _filter(self, token: SearchFilter) -> RuleCondition:
+        operator = _SEARCH_TO_RELAY_OPERATORS.get(token.operator)
+        if not operator:
+            raise ValueError(f"Unsupported operator {token.operator}")
+
+        value = token.value.raw_value
+        if operator == "eq" and isinstance(value, str) and "*" in value:
```

Comment:
I'll confirm if this check is sufficient and if there's a function independent of the query builder we can reuse for this. Please keep this comment open in the meanwhile.

Reply:
Switched to `is_wildcard()`